File: demos/showcase/android/views/src/main/java/com/bajutsu/showcase/views/Net.kt
package com.bajutsu.showcase.views

import dev.bajutsu.android.BajutsuNet
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * The showcase's network calls (SPEC §6), shared verbatim with the compose module — over OkHttp so
 * BajutsuAndroid's interceptor can observe them (BE-0283 — the Android peer of BajutsuKit's
 * `URLProtocol` on iOS). Each call mirrors its outcome to a `*.status` value (`loading` →
 * `done`/`error`) so a scenario can wait on the response before asserting. Any HTTP response counts as
 * `done`; only a transport failure is `error`.
 */
object Net {
    private val JSON = "application/json".toMediaType()

    private val client = OkHttpClient.Builder()
        // Finite timeouts: a hung endpoint must resolve to `error`, not park the status on `loading`
        // forever (the iOS twin's URLSession times out the same way).
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        // The one line that gives bajutsu network capture on Android; inert unless bajutsu injected a
        // collector (BE-0283).
        .addInterceptor(BajutsuNet.interceptor())
        .build()

    /** GET a URL, returning `done` on any response, `error` on a transport failure or timeout. */
    suspend fun get(urlString: String): String = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder()
                .url(urlString)
                .header("Authorization", "Bearer demo-secret-abc123")
                .build()
            client.newCall(req).execute().use { "done" }
        } catch (e: Exception) {
            "error"
        }
    }

    /**
     * POST the training-log entry to `<base>/post` as JSON. Carries a secret header
     * (`Authorization: Bearer …`) and a `password` body field so redaction has something to mask
     * (SPEC §6 / DESIGN §9).
     */
    suspend fun postLog(base: String, note: String, count: Int, intense: Boolean): String =
        withContext(Dispatchers.IO) {
            try {
                val payload = JSONObject()
                    .put("note", note)
                    .put("count", count)
                    .put("intense", intense)
                    .put("password", "hunter2")
                val req = Request.Builder()
                    .url("$base/post")
                    .header("Authorization", "Bearer demo-secret-abc123")
                    .post(payload.toString().toRequestBody(JSON))
                    .build()
                client.newCall(req).execute().use { "done" }
            } catch (e: Exception) {
                "error"
            }
        }
}

File: demos/showcase/flutter/lib/net.dart
import 'dart:convert';
import 'dart:io';

/// The showcase's network calls (SPEC §6). Each call mirrors its outcome to a `*.status` value
/// (`loading` → `done`/`error`) so a scenario can wait on the response before asserting. Any HTTP
/// response counts as `done`; a transport failure, or the total deadline elapsing, is `error`.
///
/// A request deliberately carries a secret header (`Authorization: Bearer …`) and, on the POST, a
/// `password` body field so redaction has something to mask (SPEC §6 / DESIGN §9).
///
/// Unlike the iOS (URLSession + BajutsuKit) and Android (OkHttp + BajutsuAndroid) twins, these use
/// Dart's own `HttpClient`, which does not flow through the native interceptors, so `network`
/// capture and `mocks` do not observe Flutter traffic (BE-0008 out-of-scope note in drivers.md).
/// The status mirror still drives the deterministic wait/assert the shared scenarios rely on.
class Net {
  // A total deadline over the whole exchange, not just the TCP connect: an endpoint that accepts the
  // connection and then never responds must still resolve to `error`, matching the URLSession /
  // OkHttp request timeouts the native twins use, rather than parking the status on `loading`.
  static const Duration _deadline = Duration(seconds: 15);

  static final HttpClient _client = HttpClient()..connectionTimeout = _deadline;

  /// GET a URL, returning `done` on any response, `error` on a transport failure or timeout.
  static Future<String> get(String url) => _run(() async {
        final request = await _client.getUrl(Uri.parse(url));
        request.headers.set(HttpHeaders.authorizationHeader, 'Bearer demo-secret-abc123');
        final response = await request.close();
        await response.drain<void>();
      });

  /// POST the training-log entry to `<base>/post` as JSON.
  static Future<String> postLog(String base, String note, int count, bool intense) => _run(() async {
        final request = await _client.postUrl(Uri.parse('$base/post'));
        request.headers.set(HttpHeaders.authorizationHeader, 'Bearer demo-secret-abc123');
        request.headers.contentType = ContentType.json;
        request.write(jsonEncode({'note': note, 'count': count, 'intense': intense, 'password': 'hunter2'}));
        final response = await request.close();
        await response.drain<void>();
      });

  static Future<String> _run(Future<void> Function() call) async {
    try {
      await call().timeout(_deadline);
      return 'done';
    } catch (_) {
      return 'error';
    }
  }
}
